tree_segment.py

```diff
@@ -0,0 +1,434 @@
+import laspy
+import matplotlib.pyplot as plt
+import math
+import numpy as np
+import random
+import time
+import pandas as pd
+import argparse
+import re
+import os
+
+
+class Point_Cloud:
+    # object that stores the points in a point cloud, initialized from a las file
+    # points can be enabled or disabled, disabled points are ignored when searching for stems
+    def __init__(self, file_name, verbose=False):
+        all_data = laspy.read(file_name)
+        height_of_points_above_ground = np.array(
+            [point[-1] for point in all_data.points.array])
+        self.points = [Point(xyz, height) for xyz, height in zip(
+            all_data.xyz, height_of_points_above_ground)]
+        # internal, to access this use enabled_points()
+        self.__enabled_points__ = self.points
+        self.verbose = verbose
+
+    def enabled_points(self):
+        # filters the set of points for just those enabled, then returns that set of enabled points
+        self.__enabled_points__ = filter_for_enabled(self.__enabled_points__)
+        return self.__enabled_points__
+
+    def find_highest_point(self):
+        # returns the highest point in the point cloud, ie the one with the greatest z coordinate
+        return max(self.enabled_points(), key=lambda point: point.z())
+
+    def find_stem(self, min_height=.5, fail_to_find_disable_radius=.2, descendents_height=.3):
+        # returns a stem from the point cloud if it exists, otherwise returns None
+        # the stem will be a list of points in descending z value, starting from above the min height and going down to a ground point
+        # inputs:
+        #   - min_height : the minimum height (in m) for a valid stem
+        #   - fail_to_find_disable_radius : when a point has no descendents, we disable each point within this radius of it
+        #   - descendents_height : the maximum vertical distance allowed to a descendent of this point
+        stem_grounded = False
+        while not(stem_grounded):
```

Comment:
Can this block be re-structured to reduce the levels of nesting/complexity required?
e.g.
```
stem = []
while True:
    if not stem:
        highest_point = self.find_highest_point()
        if highest_point.height_above_ground < min_height:
            return None
         stem = [highest_point]
    // [if we reach this point, stem[-1] is valid, one way or another]
    active_point = stem[-1]
    descendants = active_point.find_descendants(...)
    if not descendants:
         self.disable_stem_region(...)
         stem.pop()
         continue  // skip over the rest of the loop
     // [if we reach this point, stem[-1] has a valid descendant]
     next_point = random.choice(descendants)
     stem.push(next_point)
     if next_point.is_ground:
         return stem
```

Meta Commentary:
Reasoning about this is still somewhat non-trivial, and might benefit from further changes. But reducing nesting depth is often low-hanging fruit for readability, there are a few common tactics which can be implemented fairly trivially:
* Check the assumptions you need first, and either fix them or (more often) return if they can't be satisfied.
* Move nested blocks into individual functions.
There's a really great video on this topic: https://youtu.be/CFRhGnuXG-4 (9 minutes)

More generally, if you can identify the right data structure you can often structure your code around that, e.g. this function is equivalent to depth-first search over the DAG of all points connected by the descendant relation (with one artificial point at the top connecting to every point). This helped convince me that a single while loop was sufficient, but is also a more general reasoning tool. e.g. when thinking about which descendants to explore first, or determining whether the algorithm could be sped up by pre-computing the descendants relationship.

Reply:
Good idea! I did a major refactor along these lines, and I think the code is far more readable now!